pkgs/development/libraries/science/math/cudnn/generic.nix

```diff
@@ -17,44 +22,48 @@
 , url
 , hash ? null
 , sha256 ? null
-, supportedCudaVersions ? []
+, supportedCudaVersions ? [ ]
 }:
 
 assert (hash != null) || (sha256 != null);
 
 let
+  inherit (cudatoolkit) cc;
+
   majorMinorPatch = version: lib.concatStringsSep "." (lib.take 3 (lib.splitVersion version));
   version = majorMinorPatch fullVersion;
-in stdenv.mkDerivation {
-  name = "cudatoolkit-${cudatoolkit.majorVersion}-cudnn-${version}";
+
+  useRedist = libcublas != null;
+  cudaDeps = if useRedist then [ libcublas ] else [ cudatoolkit ];
+in
+stdenv.mkDerivation {
+  name = "cudatoolkit-${cudaMajorVersion}-cudnn-${version}";
 
   inherit version;
-  # It's often the case that the src depends on the version of cudatoolkit it's
-  # being linked against, so we pass in `cudatoolkit` as an argument to `mkSrc`.
   src = fetchurl {
     inherit url hash sha256;
   };
 
-  nativeBuildInputs = [ addOpenGLRunpath ];
+  # Check and normalize Runpath against DT_NEEDED using autoPatchelf.
+  # Prepend /run/opengl-driver/lib using addOpenGLRunpath for dlopen("libcudacuda.so")
+  nativeBuildInputs = [
+    autoPatchelfHook
+    autoAddOpenGLRunpathHook
+  ];
 
-  # Some cuDNN libraries depend on things in cudatoolkit, eg.
-  # libcudnn_ops_infer.so.8 tries to load libcublas.so.11. So we need to patch
-  # cudatoolkit into RPATH. See also https://github.com/NixOS/nixpkgs/blob/88a2ad974692a5c3638fcdc2c772e5770f3f7b21/pkgs/development/python-modules/jaxlib/bin.nix#L78-L98.
+  # Used by autoPatchelfHook
+  buildInputs = builtins.map lib.getLib [
```

Comment:
Isn't that done automatically?

Reply:
libcudnn_ops_infer.so.8 may not be in DT_NEEDED/picked up by autoPatchelfHook.

Reply:
If you mean `getLib` - I think it ends out extracting the default outputs here, I automatically just repeated the pattern I saw somewhere in nixpkgs.
If the question is about the doings of autoPatchelfHook, then yes, there are two factors at play:

1. `autoPatchelf` needs to know in which paths to look for the libraries it finds missing, so we have tell about `gcc.cc`, etc
2. Samuel is correct, `libcudnn_ops_infer.so.8` is not part of NEEDED, which is why we also patch it below (that is more of a temporary measure though)